### PR TITLE
[STJ SourceGen] Use Roslyn4.4 target when testing

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Created this PR to try to isolate the test failures reported in https://github.com/dotnet/runtime/pull/79828#issuecomment-1362096784

### CODE REVIEW

**Review by ViktorHofer — APPROVED** [review]
I'm OK with not testing the roslyn4.0 source generator for now. We do the same for target frameworks and only test the very latest. If we notice regressions in the roslyn4.0 source generator though (i.e. based on customer reports), we should invest in testing it as well.

**Comment by eiriktsarpalis** [comment]
I neglected to put in the description that I created this PR in order to isolate some wasm test failures I've been seeing: https://github.com/dotnet/runtime/pull/79828#issuecomment-1362096784

TL;DR running tests against Roslyn4.4 causes deterministic assertion failures when running release builds on wasm. I don't see how a compile time artifact such as the Roslyn version could trigger test failures at runtime, but I'll keep investigating.

**Comment by eiriktsarpalis** [comment]
Hi @dotnet/mono-team, it looks like the changes introduced in 5236bb87b3f862a6a5d52e662ab5db792215954f are causing a number of tests to fail on wasm only. I've been able to narrow down the failures to a single unit test in fe3a6270757967cd1cd9000f2b49adf0ae85e082. The failure happens deterministically and only occurs on wasm targets. It appears to occur because the `JsonDocument._utf8Json` span is reported as empty, even though my changes in the second commit make this case impossible -- I suspect this could be caused by some kind of memory corruption.
